### DISCUSSION TITLE
FASTLED Bankable?

### DISCUSSION THREAD

**Discussion by EvgeniySavkin** [tttapa/Control-Surface#401] [General]
Hi Peter!!! I still admire your work!!! And again, I can't understand the work of banks. The previous project(#379) was successful!!! But a person is always not enough)) That's not enough for me. To work, I would need a controller with the capabilities of Launchpad from the manufacturer Novation, with its multi-colored functionality. To start, I would like to understand how to access FASTLED in the NoteButton class, but since there are 64 buttons (8 by 8), and there are two banks, and I am stupid in programming, I ask you to help me master the address LEDs in the array..
```
#include <Control_Surface.h>
#include <AH/Hardware/MCP23017Encoders.hpp>
#include <Wire.h>

// Type for the MCP23017 encoders (translates encoder pulses to position)
using WireType = decltype(Wire);     // The type of I²C driver to use
using EncoderPositionType = int32_t; // The type for saving encoder positions
constexpr bool IntSafe = true;       // Make it safe to call `update` in an ISR
using MCPEncoderType = MCP23017Encoders<WireType, EncoderPositionType, IntSafe>;
 
// Type for the MIDI encoders (translates position to MIDI messages)
template <uint8_t NumBanks>
using MCPCCAbsEncType = Bankable::GenericMIDIAbsoluteEncoder<MCPEncoderType::MCP23017Encoder,    // Encoder type
                                                             NumBanks, Bankable::SingleAddress,  // Bank size and behavior
                                                             ContinuousCCSender>;                // MIDI output
// Wrapper for consistency with the standard Bankable::CCAbsoluteEncoder interface
template <uint8_t NumBanks>
struct MCPBankableCCAbsoluteEncoder : MCPCCAbsEncType<NumBanks> {
  MCPBankableCCAbsoluteEncoder(BankConfig<NumBanks> config, MCPEncoderType::MCP23017Encoder enc,
                               MIDIAddress address, int16_t multiplier = 1, uint8_t pulsesPerStep = 4)
    : MCPCCAbsEncType<NumBanks>({config, address}, std::move(enc), multiplier, pulsesPerStep, {}) {}
};

HairlessMIDI_Interface midi;
HardwareSerialMIDI_Interface serialmidi = {Serial1, MIDI_BAUD};

BidirectionalMIDI_PipeFactory<2> pipes;

const uint8_t interrupt_pin = 12;

Bank<2> bank(8);

IncrementDecrementSelector<2> selector = {
  bank,
  {10, 11},
  Wrap::Wrap,
};
//>>>>>>>>>>>
// Create an object that manages the 8 encoders connected to the MCP23017.
MCPEncoderType enc = {Wire, 0x0, interrupt_pin};
//                    │     │    └─ Interrupt pin
//                    │     └────── Address offset
//                    └──────────── I²C interface

//Enc
MCPBankableCCAbsoluteEncoder<2> ccencoders[] = {
  {{bank, BankType::CHANGE_ADDRESS}, enc[0], {65, CHANNEL_1}, 1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, enc[1], {66, CHANNEL_1}, 1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, enc[2], {67, CHANNEL_1}, 1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, enc[3], {68, CHANNEL_1}, 1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, enc[4], {69, CHANNEL_1}, 1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, enc[5], {70, CHANNEL_1}, 1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, enc[6], {71, CHANNEL_1}, 1, 4},
  {{bank, BankType::CHANGE_ADDRESS}, enc[7], {72, CHANNEL_1}, 1, 4},
};
//>>>>>>>>
//RingsEnc
SPIShiftRegisterOut<8 * 16> sreg = {
  53,
  MSBFIRST,
};

Bankable::CCLEDBar<2, 16> leds[] {
  {{bank, BankType::CHANGE_ADDRESS}, sreg.pins().slice<0, 15>(), {0x00, CHANNEL_1}}, //The LED ring displays the value Channel_01
  {{bank, BankType::CHANGE_ADDRESS}, sreg.pins().slice<16, 31>(), {0x01, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, sreg.pins().slice<32, 47>(), {0x02, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, sreg.pins().slice<48, 63>(), {0x03, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, sreg.pins().slice<64, 79>(), {0x04, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, sreg.pins().slice<80, 95>(), {0x05, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, sreg.pins().slice<96, 111>(), {0x06, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, sreg.pins().slice<112, 127>(), {0x07, CHANNEL_1}},
};
//>>>>>>>>>>
//ButtonMUX
CD74HC4067 mux1 = {
  2,               // input pin
  {3, 4, 5, 6}, // Address pins S0, S1, S2, S3
  // 7, // Optionally, specify the enable pin
};

Bankable::CCButtonLatched<2> button1[] = {
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(0), {1, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(1), {2, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(2), {3, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(3), {4, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(4), {5, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(5), {6, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(6), {7, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(7), {8, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(8), {9, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(9), {10, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(10), {11, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(11), {12, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(12), {13, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(13), {14, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(14), {15, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux1.pin(15), {16, CHANNEL_1}},
};

CD74HC4067 mux2 = {
  3,               // input pin
  {3, 4, 5, 6}, // Address pins S0, S1, S2, S3
  // 7, // Optionally, specify the enable pin
};

Bankable::CCButtonLatched<2> button2[] = {
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(0), {17, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(1), {18, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(2), {19, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(3), {20, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(4), {21, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(5), {22, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(6), {23, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(7), {24, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(8), {25, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(9), {26, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(10), {27, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(11), {28, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(12), {29, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(13), {30, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(14), {31, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux2.pin(15), {32, CHANNEL_1}},
};

CD74HC4067 mux3 = {
  4,               // input pin
  {3, 4, 5, 6}, // Address pins S0, S1, S2, S3
  // 7, // Optionally, specify the enable pin
};

Bankable::CCButtonLatched<2> button3[] = {
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(0), {33, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(1), {34, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(2), {35, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(3), {36, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(4), {37, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(5), {38, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(6), {39, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(7), {40, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(8), {41, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(9), {42, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(10), {43, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(11), {44, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(12), {45, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(13), {46, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(14), {47, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux3.pin(15), {48, CHANNEL_1}},
};

CD74HC4067 mux4 = {
  5,               // input pin
  {3, 4, 5, 6}, // Address pins S0, S1, S2, S3
  // 7, // Optionally, specify the enable pin
};

Bankable::CCButtonLatched<2> button4[] = {
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(0), {49, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(1), {50, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(2), {51, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(3), {52, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(4), {53, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(5), {54, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(6), {55, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(7), {56, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(8), {57, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(9), {58, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(10), {59, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(11), {60, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(12), {61, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(13), {62, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(14), {63, CHANNEL_1}},
  {{bank, BankType::CHANGE_ADDRESS}, mux4.pin(15), {64, CHANNEL_1}},
};
//>>>>>>>>>>> 
void isr() {
  enc.update(); // Read the state of the encoders and update the positions
}



void setup() {
    // Manually connect the MIDI interfaces to Control Surface
  Control_Surface | pipes | midi;
  Control_Surface | pipes | serialmidi;
  Control_Surface.begin();
  Wire.begin();
  Wire.setClock(800000);
  enc.begin();
  attachInterrupt(digitalPinToInterrupt(interrupt_pin), isr, LOW);
}

void loop() {
  // No `enc.update()` here, updating the encoders happens inside of the ISR
  Control_Surface.loop();
```
Please do not pay attention to classes, addresses, and pins. Hardware Atmega2560 configuration, new-input branch, and SK6812 2020 LEDs(will they work??)

**Comment by tttapa** [comment]
Please post the minimal amount of code needed to answer your question. It's not clear to me what you're asking. The NoteButton class is independent from any (Fast)LEDs you may be using. When a button is pressed, a MIDI message is sent to the software on your computer, that software registers the button press, decides which LEDs should be updated, and send a MIDI message back to the Arduino containing updated LED color values. The buttons don't control the LEDs directly, the software on your computer does.

Have you tried the FastLED example? https://tttapa.github.io/Control-Surface-doc/new-input/Doxygen/d8/dd2/9_8Note-FastLED_8ino-example.html  
It uses the Novation Launchpad color scheme.

There is no bankable NoteRangeFastLED class yet, because I haven't had the time to write it. It shouldn't be too hard, most of it can be copied from the non-bankable version. If you're certain that you need a Bankable::NoteRangeFastLED, I'll see if I can add it to the library.

What is the relationship between the 2 banks and the 8×8 LED matrix? How much LEDs do you intend to wire up? All 64? Just 32 in a configuration to switch between two banks?

**Reply by EvgeniySavkin** [reply]
>  When a button is pressed, a MIDI message is sent to the software on your computer, that software registers the button press, decides which LEDs should be updated, and send a MIDI message back to the Arduino containing updated LED color values. The buttons don't control the LEDs directly, the software on your computer does.

I understand that the LEDs are independent of the buttons. I don't understand how to use the example: https://tttapa.github.io/Control-Surface-doc/new-input/Doxygen/d8/dd2/9_8Note-FastLED_8ino-example.html

In these lines, the LEDs are advertised:
```
// Define the array of leds.
Array<CRGB, 8> leds = {};
```
And instead of the number " 8 " you need to put "64"???
/////////////////////

```
// The data pin with the strip connected.
constexpr uint8_t ledpin = 2;
```
Nothing changes in this line, we initialize the PIN for these LEDs with this code.
That is, I understand correctly that you can connect all 64 LEDs to 2 pins???
/////////////////////
```
// Create a MIDI input element that listens to all notes in the range C4 - G4
// (the range starts at C4 and has a length equal to `leds.length` == 8).
NoteRangeFastLED<leds.length> midiled = {leds, note(C, 4)};
```
In this line of code, I do not understand how to register all 64 LEDs correctly.
That is, the top panel of the controller will have all 64 buttons and 64 RGB LEDs.

**Reply by EvgeniySavkin** [reply]
> What is the relationship between the 2 banks and the 8×8 LED matrix? How much LEDs do you intend to wire up? All 64? Just 32 in a configuration to switch between two banks?

Without banks, the controller will be limited in the number of channels. It would also be nice to prescribe switching both on one channel and on eight channels at a time. But not necessarily. Thank you again for your hard work and knowledge. Hello from Russia.

**Reply by tttapa** [reply]
It might be worth having a look at the examples of the [FastLED library](https://github.com/FastLED/FastLED) first.

Basically, you have an array of type CRGB (Color red/green/blue). Each element of the array corresponds to one LED, so the size of the array (8 in the example) should correspond to the number of LEDs you have (64 in your case).  
When you change the CRGB value of element `leds[3]`, this will eventually cause the fourth LED to change to that color.  
That's exactly what the `NoteRangeFastLED` class does: it writes color values to the `leds` array when it receives certain MIDI messages.  
It notifies you when it has changed any color values through the `getDirty()` method. When this method returns true, colors were updated and you should push the `leds` array to the actual LEDs using `FastLED.show()`. (This is the same as in the FastLED library examples.)

> That is, I understand correctly that you can connect all 64 LEDs to 2 pins???

The `ledpin` constant is the Arduino pin number connected to the data input of the LEDs. All LEDs are driven from a single data pin (and power and ground, make sure you use a beefy power supply if you want to use 64 LEDs). 

> And instead of the number " 8 " you need to put "64"???

Yes.

> In this line of code, I do not understand how to register all 64 LEDs correctly.

If you changed the length of the `leds` array to 64, you don't have to change anything here, since this line uses `leds.length` as an argument, i.e. the lenght of the `leds` array.

> Without banks, the controller will be limited in the number of channels. It would also be nice to prescribe switching both on one channel and on eight channels at a time.

If I understand correctly you want to have 64 LEDs times 16 channels and be able to cycle through the 16 channels? You'll need quite a bit of RAM (3×64 bytes for the LED color buffer and 16×64 bytes for the saved MIDI data), that's well over a KiB. It should be possible, though.

**Reply by EvgeniySavkin** [reply]
> If I understand correctly you want to have 64 LEDs times 16 channels and be able to cycle through the 16 channels?

I want to use 8 channels, with 8 buttons. 
This will be a copy of the Launchpad pro controller 
https://novationmusic.com/en/launch/launchpad-pro
Although not with the full implementation of its functions.

**Reply by tttapa** [reply]
How many LEDs? The Launchpad Pro has 64. Do you want to have 8 physical LEDs to display all 64 Launchpad LEDs using 8 banks?

**Reply by EvgeniySavkin** [reply]
No, no! I need 64 LEDs as on Launchpad. And just 2 banks is enough for me.

**Reply by EvgeniySavkin** [reply]
There was a thought or even a question- 
Is it possible to manually add addresses for LEDs???

**Reply by tttapa** [reply]
> No, no! I need 64 LEDs as on Launchpad. And just 2 banks is enough for me.

It's still not clear to me what you actually want to do. You want 8 buttons but 64 LEDs? What should the banks do?

> Is it possible to manually add addresses for LEDs???

Yes, but you usually want to avoid that for performance reasons. You could use the index permuter if the MIDI notes aren't in the same order of the LEDs, see https://tttapa.github.io/Control-Surface-doc/new-input/Doxygen/d3/d98/classNoteCCKPRangeFastLED.html#a353b91dbf60c992aa7de7861136f339d

**Reply by tttapa** [reply]
Something to keep in mind is that driving 64 LEDs will turn off interrupts for a significant amount of time, possibly causing loss of MIDI data, see https://github.com/FastLED/FastLED/wiki/Interrupt-problems

**Reply by EvgeniySavkin** [reply]
> It's still not clear to me what you actually want to do. You want 8 buttons but 64 LEDs? What should the banks do?

I want 64 buttons, with 64 LEDs. And 2 banks for my tasks will be enough.
![Launchpad](https://user-images.githubusercontent.com/45334528/108639327-c0f33900-74b5-11eb-8169-a40709c22a71.png)

> Yes, but you usually want to avoid that for performance reasons. You could use the index permuter if the MIDI notes aren't in the same order of the LEDs, see https://tttapa.github.io/Control-Surface-doc/new-input/Doxygen/d3/d98/classNoteCCKPRangeFastLED.html#a353b91dbf60c992aa7de7861136f339d

I'm asking for the future. If I run into difficulties in a subsequent solution, it will be possible to solve the problem. Also, judging by the instructions for Launchpad, I saw that the controller has a mode with different midi maps, including custom.

I apologize very much. It is very difficult to formulate a thought so that GoogleTranslate translates correctly.

**Reply by EvgeniySavkin** [reply]
> Something to keep in mind is that driving 64 LEDs will turn off interrupts for a significant amount of time, possibly causing loss of MIDI data, see https://github.com/FastLED/FastLED/wiki/Interrupt-problems

I use LEDs with only one data line. That is-pin " Data In, Vdd, Data Out and GND"


P.S.: I guess I'm confused. Solution to the problem with interrupts, use the APA102 LEDs? That is, with the Data line and CLK?

**Reply by EvgeniySavkin** [reply]
Hi Peter! I'll start the question again!
How to create a Bank for LEDs, for example, as with buttons (Bankable:: NoteButton) ?
```
#include <FastLED.h>
// Must be before Control Surface to enable FastLED features of Control Surface
#include <Control_Surface.h>
// Define the array of leds.
Array<CRGB, 16> fleds = {};

constexpr uint8_t ledpin = 10;

HairlessMIDI_Interface midi;

Bank<2> bank(8);
IncrementDecrementSelector<2> selector = {
  bank,
  {10, 11},
  Wrap::Wrap,
};

using namespace MIDI_Notes;
//How create Bankable for FastLED???
NoteRangeFastLED<fleds.length> midiled {fleds, MIDI_Notes::C(2)};

CD74HC4067 mux = {
  2,               // input pin
  {3, 4, 5, 6}, // Address pins S0, S1, S2, S3
  // 7, // Optionally, specify the enable pin
};

Bankable::NoteButton button[] = {
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(0), MIDI_Notes::C(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(1), MIDI_Notes::Db(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(2), MIDI_Notes::D(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(3), MIDI_Notes::Eb(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(4), MIDI_Notes::E(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(5), MIDI_Notes::F_(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(6), MIDI_Notes::Gb(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(7), MIDI_Notes::G(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(8), MIDI_Notes::Ab(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(9), MIDI_Notes::A(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(10), MIDI_Notes::Bb(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(11), MIDI_Notes::B(2)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(12), MIDI_Notes::C(3)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(13), MIDI_Notes::Db(3)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(14), MIDI_Notes::D(3)},
  {{bank, BankType::CHANGE_ADDRESS}, mux.pin(15), MIDI_Notes::Eb(3)},
};

void setup() {
  Control_Surface.begin(); // Initialize Control Surface
  // See FastLED examples and documentation for more information.
  FastLED.addLeds<NEOPIXEL, ledpin>(fleds.data, fleds.length);
  FastLED.setCorrection(TypicalPixelString);
  midiled.setBrightness(128);
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface
  FastLED.show();
}
```

**Reply by tttapa** [reply]
I've pushed an implementation of `Bankable::NoteRangeFastLED` to the [`bankable-fastled`](https://github.com/tttapa/Control-Surface/tree/bankable-fastled) branch. 

Example code:
```cpp
#include <FastLED.h>
// Must be before Control Surface to enable FastLED features of Control Surface
#include <Control_Surface.h>

// Define the array of leds.
Array<CRGB, 8> leds {};
// The data pin with the strip connected.
constexpr uint8_t ledpin = 13;

USBMIDI_Interface midi;

constexpr uint8_t NBanks = 3;
Bank<NBanks> bank;
IncrementSelector<NBanks> sel {bank, 2};

// Create a functor that maps the velocity and the index of a note to a color.
struct RainbowColorMapper {
  CHSV operator()(uint8_t velocity, uint8_t bankIndex, uint8_t index) const {
    uint8_t h = 255 * (index + bankIndex * leds.length) / (leds.length * NBanks);
    return CHSV(h, 255, 255u * velocity / 127u);
  }
};

Bankable::NoteRangeFastLED<NBanks, leds.length, RainbowColorMapper> midiled {
  {bank, BankType::CHANGE_CHANNEL},
  leds,
  MIDI_Notes::C(2),
};

void setup() {
  // See FastLED examples and documentation for more information.
  FastLED.addLeds<NEOPIXEL, ledpin>(leds.data, leds.length);
  FastLED.setCorrection(TypicalPixelString);
  midiled.setBrightness(128);
  Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
  if (midiled.getDirty()) { // If the colors changed
    FastLED.show();         // Update the LEDs with the new colors
    midiled.clearDirty();   // Clear the dirty flag
  }
}
```
As you might imagine, it uses quite a bit of memory.

I haven't generated the documentation for this branch, if you have any questions, let me know.
